### PR TITLE
[MetaSchedule] XGB-based Cost Model

### PR DESCRIPTION
**Pull request by junrushao1994**
This PR is part of the stage M3c of the meta schedule project (#8473).

The architecture is re-designed by Junru and Xiyou. In this PR we introduced a XGB-based cost model based on meta schedule's cost model interface. Unittests are included.

Thanks to all co-authors for contributing!

Co-authored-by: Xiyou Zhou \<xiyou@octoml.ai\>
Co-authored-by: Bohan Hou \<32121147+spectrometerHBH@users.noreply.github.com\>
Co-authored-by: Ruihang Lai \<lairuihangdongdong@qq.com\>
Co-authored-by: Hongyi Jin \<3231950289@qq.com\>
Co-authored-by: Wuwei Lin \<wuwei@apache.org\>
Co-authored-by: Siyuan Feng \<Hzfengsy@sjtu.edu.cn\>

### CODE REVIEW

**Comment by junrushao1994** [comment]
CC: @comaniac @yzhliu @merrymercy

File: python/tvm/meta_schedule/cost_model/xgb_model.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+
+        Parameters
+        ----------
+        path : str
+            The file path.
+
+        Note
+        ----
+        Since XGBoost model trains from scratch, each time we can only load the model without the
+        previous cached features / results so any call of update won't use previous training data.
```
Comment (comaniac) [inline python/tvm/meta_schedule/cost_model/xgb_model.py:320]: I didn't quite get this note. It seems to me that you load the previous model along with cached features and costs. Doesn't the loaded cached features and costs used in future training?

Comment (junrushao1994) [reply python/tvm/meta_schedule/cost_model/xgb_model.py:320]: Sorry, this comment is wrong (CC @zxybazh). Actually we did load/store the cached feature vectors to make sure all the existing data are used for training.

Comment (junrushao1994) [reply python/tvm/meta_schedule/cost_model/xgb_model.py]: Updated with accurate documentation

Comment (zxybazh) [reply python/tvm/meta_schedule/cost_model/xgb_model.py]: Thanks for point this out. This comment was made before we store the cached features so please feel free to remove it.

**Comment by junrushao1994** [comment]
@comaniac Thanks for the extremely valuable feedback!

> when training data gets bigger and bigger, the time to train the XGBoost cost model becomes tedious even the accuracy isn't further improved

That's exactly what I'm observing too! In this particular case, hyper-parameters of XGB might not be suitable any more, which limits the model capacity, and we might have to tweak around to find out the best hyperparameters.

> What Ansor has done is simply reduce the re-training frequency (e.g., re-train per 2 rounds) when training data size is larger than a threshold.

This is how Ansor deals with this right now...We might consider better heuristics in the future, including switching models, tweaking model capacity with AutoML stuff, etc.

> we can also refer to the accuracy between the predicted cost and new measured latencies to determine whether to re-train the model in the next round

Using our current interface, this is pretty simple to do so. We have a `validate` method that allows us to validate the rmse of the cost model's prediction - and I used this method quite frequently in model debugging too.


Anyway, I think we are pretty aligned with the methodology and path to improvement. Let's work together to improve it in the future

**Review by zxybazh — APPROVED** [review]
LGTM.
